File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
from collections import defaultdict, deque


def bfs(tree, n, v1, v2):
    queue = deque([v1])
    
    visited = [False] * (n + 1)
    visited[v1], visited[v2] = True, True
    
    cnt = 1
    
    while queue:
        v = queue.popleft()
        
        for adj_v in tree[v]:
            if not visited[adj_v]:
                queue.append(adj_v)
                visited[adj_v] = True
                cnt += 1
                
    return cnt
            

def solution(n, wires):
    answer = float('inf')
    
    tree = defaultdict(list)
    
    for v1, v2 in wires:
        tree[v1].append(v2)
        tree[v2].append(v1)
        
    cnt = 0
        
    for v1 in tree.keys():
        for v2 in tree[v1]:
            cnt = bfs(tree, n, v1, v2)
            
            answer = min(answer, abs(n - 2 * cnt))
            
    return answer

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.kt
import kotlin.math.abs

class Solution {
    fun solution(n: Int, wires: Array<IntArray>): Int {
        var answer = mutableListOf<Int>()
        
        for (removedEdge in wires.indices) {
            val tree = mutableMapOf<Int, MutableList<Int>>().withDefault {mutableListOf()}
            
            for ((edge, wire) in wires.withIndex()) {
                if (removedEdge == edge) continue
                
                tree.getOrPut(wire[0]) {mutableListOf()}.add(wire[1])
                tree.getOrPut(wire[1]) {mutableListOf()}.add(wire[0])
            }
            
            val queue = ArrayDeque<Int>()
            queue.addLast(1)
            
            val visited = mutableSetOf<Int>()
            visited.add(1)
            
            var towerCnt = 1
            
            while (queue.isNotEmpty()) {
                val currTower = queue.removeFirst()
                
                for (adjTower in tree.getValue(currTower)) {
                    if (adjTower !in visited) {
                        queue.addLast(adjTower)
                        visited.add(adjTower)
                        towerCnt++
                    }
                }
            }
            
            answer.add(abs(n - 2 * towerCnt))
        }
        
        return answer.minOrNull() ?: 0
    }
}